src/HeroOptionsComponent.js

```diff
@@ -0,0 +1,117 @@
+import React, {Component} from 'react';
+import {Radio} from './RadioButtonCatalog';
+import {withFormField} from './FormField';
+import queryString from 'query-string';
+import {MDCTextField} from '@material/textfield/index';
+
+export class HeroOptionsComponent extends Component {
+  render() {
+    return (
+        <div className='hero-options'>
+          <ul className='mdc-list mdc-list--non-interactive'>
+            {this.props.config.options && this.props.config.options.map((option, index) => {
+              return <Option key={option.name + index} option={option} {...this.props}/>
+            })}
+          </ul>
+        </div>
+    );
+  }
+}
+
+const Option = (props) => {
+  return getOptionComponent(props.option, props);
+};
+
+const getOptionComponent = (option, props) => {
+  switch(option.type) {
+    case 'label':
+      return <li className='mdc-list-item'><LabelOption>{option.name}</LabelOption></li>;
+    case 'radiogroup':
+      return <RadioGroupOption {...option} {...props}/>;
+    default: // Text field
+      return <TextFieldOption {...option} {...props}/>;
+  }
+};
+
+
+const RadioFormField = withFormField(Radio);
+
+const updateUrl = (history, urlParams = {}, key, newValue) => {
+  urlParams[key] = newValue;
+
+  history.push({
+    pathname: history.location.pathname,
+    search: queryString.stringify(urlParams),
+  });
+};
+
+const LabelOption = ({children}) => {
+  return (
+      <span className='mdc-typography--subtitle2'>{children}</span>
+  )
+};
+
+//TODO: Refactor to remove anon function from onChange.
+const RadioGroupOption = ({name, options, value, history, urlParams}) => {
+  return (
+      <React.Fragment>
+        <li className='mdc-list-item'><span
+            className='mdc-typography--subtitle1'>{name}</span></li>
+        {options.map((opt) => {
+          return (
+              <li key={opt.value} className='mdc-list-item'>
+                <RadioFormField
+                    name='demo-radio-set'
+                    label={opt.label}
+                    defaultChecked={value === opt.value}
+                    onChange={() => updateUrl(history, urlParams, 'type', opt.value)}/>
+              </li>
+          )
+        })}
+      </React.Fragment>
+  )
+};
+
+class TextFieldOption extends Component {
+  state = {value: ''};
+  textField = null;
+  tfRef = React.createRef();
+
+  componentDidUpdate() {
+    if (this.textField) {
+      this.textField.layout();
+    } else {
+      this.textField = new MDCTextField(this.tfRef.current);
+    }
+  }
+
+  componentDidMount() {
+    this.setState({value: this.props.value});
```

Comment:
Is there a reason to do this here rather than doing `this.state = {value: this.props.value}` in constructor?

Reply:
We can't call `setState` on a component that isn't mounted.

Reply:
I'm not talking about calling setState, I'm talking about assigning it directly. My suggestion comes straight from React's [docs on state & lifecycle](https://reactjs.org/docs/state-and-lifecycle.html#adding-local-state-to-a-class). I'm wondering if there's a specific need to wait until it's mounted, or whether we can actually just initialize the state in the constructor.